### PR TITLE
Feature-32 Add create project

### PR DESCRIPTION
**Pull request by pedro-lima-jalafund**
[AB#186](https://dev.azure.com/funjaladev/DevLevel3/_boards/board/t/DevLevel3%20Team/Stories/?workitem=186)
[AB#32](https://dev.azure.com/funjaladev/7867a1c5-5ac7-4c7f-baff-87d740bf4fc1/_workitems/edit/32)
Add Create new Project

### CODE REVIEW

**Comment by ela-ocana-jalafund** [comment]
@cristian-chavez-jalafund 
https://github.com/fundev3/teamup-projects/tree/feature/32-create-new-project

**Comment by martin-callejas-jalafund** [comment]
please resolve merge conflicts locally and push again

File: src/api/Jalasoft.TeamUp.Projects.API/Controllers/PostProject.cs [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+        {
+            this.postProjectService = postProjectService;
+        }
+
+        [FunctionName("PostProject")]
+        [OpenApiOperation(operationId: "Post", tags: new[] { "CreateProject" })]
+        [OpenApiRequestBody("application/json", typeof(Project), Description = "JSON request body containing { ID, Name, Description, Contact }")]
+        [OpenApiResponseWithBody(statusCode: HttpStatusCode.OK, contentType: "application/json", bodyType: typeof(Project), Description = "Successful response")]
+        public async Task<IActionResult> Run(
+            [HttpTrigger(AuthorizationLevel.Anonymous, "post", Route = null)] HttpRequest req,
```
Comment (rodrigo-baldiviezo-jalafund) [inline src/api/Jalasoft.TeamUp.Projects.API/Controllers/PostProject.cs]: Change null to "projects"

**Review by niver-copa-jalafund — DISMISSED** [review]

**Comment by martin-callejas-jalafund** [comment]
pr build is failing, please fix stylecop issues

File: src/api/Jalasoft.TeamUp.Projects.API/Controllers/PostProject.cs [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+    public class PostProject
+    {
+        private readonly IProjectsService postProjectService;
+
+        public PostProject(IProjectsService postProjectService)
+        {
+            this.postProjectService = postProjectService;
+        }
+
+        [FunctionName("project")]
```
Comment (cristian-chavez-jalafund) [inline src/api/Jalasoft.TeamUp.Projects.API/Controllers/PostProject.cs]:
```suggestion
        [FunctionName("PostProject")]
```

File: src/api/Jalasoft.TeamUp.Projects.API/Controllers/PostProject.cs [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+        [OpenApiOperation(operationId: "createProject", tags: new[] { "CreateProject" })]
```
Comment (cristian-chavez-jalafund) [inline src/api/Jalasoft.TeamUp.Projects.API/Controllers/PostProject.cs]:
```suggestion
        [OpenApiOperation(operationId: "PostProject", tags: new[] { "CreateProject" })]
```
just to align to the other endpoints

File: src/api/Jalasoft.TeamUp.Projects.API/Controllers/PostProject.cs [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+        [OpenApiRequestBody("application/json", typeof(Project), Description = "JSON request body")]
+        [OpenApiResponseWithBody(statusCode: HttpStatusCode.OK, contentType: "application/json", bodyType: typeof(Project), Description = "Successful response")]
```
Comment (cristian-chavez-jalafund) [inline src/api/Jalasoft.TeamUp.Projects.API/Controllers/PostProject.cs]: what is the proper http response code we should return?

File: src/dal/Jalasoft.TeamUp.Projects.DAL/ProjectsRepository.cs [hunk 5]
```diff
@@ -58,12 +58,20 @@ public class ProjectsRepository : IProjectsRepository
                 }
             };
 
+        private static IEnumerable<Project> projects = new List<Project>();
```
Comment (cristian-chavez-jalafund) [inline src/dal/Jalasoft.TeamUp.Projects.DAL/ProjectsRepository.cs]: please remove this and use the line 11 field, make the necessary changes to it ;)

**Review by cristian-chavez-jalafund — CHANGES_REQUESTED** [review]

**Comment by pedro-lima-jalafund** [comment]
[AB#173](https://dev.azure.com/funjaladev/7867a1c5-5ac7-4c7f-baff-87d740bf4fc1/_workitems/edit/173) SPIKE Task (async await) implementation with unit test

**Review by rodrigo-baldiviezo-jalafund — DISMISSED** [review]

**Review by ela-ocana-jalafund — DISMISSED** [review]

**Review by rodrigo-baldiviezo-jalafund — DISMISSED** [review]

**Review by alejandro-castellon — DISMISSED** [review]

File: src/api/Jalasoft.TeamUp.Projects.API/Controllers/PostProject.cs (hunk 2 above)
Comment (pedro-lima-jalafund) [reply src/api/Jalasoft.TeamUp.Projects.API/Controllers/PostProject.cs]: done

File: src/api/Jalasoft.TeamUp.Projects.API/Controllers/PostProject.cs (hunk 3 above)
Comment (pedro-lima-jalafund) [reply src/api/Jalasoft.TeamUp.Projects.API/Controllers/PostProject.cs]: done

File: src/api/Jalasoft.TeamUp.Projects.API/Controllers/PostProject.cs (hunk 4 above)
Comment (pedro-lima-jalafund) [reply src/api/Jalasoft.TeamUp.Projects.API/Controllers/PostProject.cs]: done

File: src/dal/Jalasoft.TeamUp.Projects.DAL/ProjectsRepository.cs (hunk 5 above)
Comment (pedro-lima-jalafund) [reply src/dal/Jalasoft.TeamUp.Projects.DAL/ProjectsRepository.cs]: done

**Review by ela-ocana-jalafund — DISMISSED** [review]

**Review by lina-ramirez-jalafund — DISMISSED** [review]

File: src/api/Jalasoft.TeamUp.Projects.API/Controllers/PostProject.cs [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+    {
+        private readonly IProjectsService postProjectService;
+
+        public PostProject(IProjectsService postProjectService)
+        {
+            this.postProjectService = postProjectService;
+        }
+
+        [FunctionName("PostProject")]
+        [OpenApiOperation(operationId: "PostProject", tags: new[] { "CreateProject" })]
```
Comment (cristian-chavez-jalafund) [inline src/api/Jalasoft.TeamUp.Projects.API/Controllers/PostProject.cs]:
```suggestion
        [OpenApiOperation(operationId: "PostProject", tags: new[] { "Projects" })]
```
to group them under Projects in the documentation

File: src/api/Jalasoft.TeamUp.Projects.API/local.settings.json [hunk 7]
```diff
@@ -5,6 +5,7 @@
     "FUNCTIONS_WORKER_RUNTIME": "dotnet"
   },
   "Host": {
+    "LocalHttpPort": 7072,
```
Comment (cristian-chavez-jalafund) [inline src/api/Jalasoft.TeamUp.Projects.API/local.settings.json:8]: I think this was for your dev testing, please could you undo this change?

File: src/dal/Jalasoft.TeamUp.Projects.DAL/ProjectsRepository.cs [hunk 8]
```diff
@@ -64,6 +64,12 @@ public Project GetProject(Guid id)
             return result;
         }
 
+        public Project PostProject(Project project)
+        {
+            Projects.Append(project);
```
Comment (cristian-chavez-jalafund) [inline src/dal/Jalasoft.TeamUp.Projects.DAL/ProjectsRepository.cs]: does this work if the Projects field is readonly (line 11) ?

File: src/api/Jalasoft.TeamUp.Projects.API/Controllers/PostProject.cs [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+        [OpenApiRequestBody("application/json", typeof(Project), Description = "JSON request body")]
+        [OpenApiResponseWithBody(statusCode: HttpStatusCode.Created, contentType: "application/json", bodyType: typeof(Project), Description = "Successful response")]
+        public async Task<IActionResult> CreateProject(
+            [HttpTrigger(AuthorizationLevel.Anonymous, "post", Route = "v1/projects")] HttpRequest req)
+        {
+            string requestBody = await new StreamReader(req.Body).ReadToEndAsync();
+            var data = JsonConvert.DeserializeObject<Project>(requestBody);
+            var result = this.postProjectService.PostProject(data);
+            return new OkObjectResult(result);
```
Comment (cristian-chavez-jalafund) [inline src/api/Jalasoft.TeamUp.Projects.API/Controllers/PostProject.cs]: in line 28 the documentation indicates a different http response code, which is the correct?

File: tests/api/Jalasoft.TeamUp.Projects.API.Tests/PostProjectsTest.cs [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+            this.mockProjectsService = new Mock<IProjectsService>();
+            this.mockHttpContext = new DefaultHttpContext();
+            this.postProject = new PostProject(this.mockProjectsService.Object);
+        }
+
+        [Fact]
+        public async void PostProject_Returns_OkObjectResult_Project()
+        {
+            var request = this.mockHttpContext.Request;
+            this.mockProjectsService.Setup(service => service.PostProject(null)).Returns(new Project());
```
Comment (cristian-chavez-jalafund) [inline tests/api/Jalasoft.TeamUp.Projects.API.Tests/PostProjectsTest.cs]: please instead of null, could you set up something like this to setup the body
public static Stream SetStream(string body)
        {
            var stream = new MemoryStream();
            var writer = new StreamWriter(stream);
            writer.Write(body);
            writer.Flush();
            stream.Position = 0;
            return stream;
        }

that will be called in your test

HttpRequest whatever = ...;
            whatever.Headers.Add("somename", "somevalue");
            whatever.Body = SetStream(
@"{ ""mybodyproperty"": ""myvalueproperty""}"
)

File: src/dal/Jalasoft.TeamUp.Projects.DAL/ProjectsRepository.cs (hunk 8 above)
Comment (pedro-lima-jalafund) [reply src/dal/Jalasoft.TeamUp.Projects.DAL/ProjectsRepository.cs]: sorry, my bad

File: src/api/Jalasoft.TeamUp.Projects.API/Controllers/PostProject.cs (hunk 9 above)
Comment (pedro-lima-jalafund) [reply src/api/Jalasoft.TeamUp.Projects.API/Controllers/PostProject.cs]: 201 of course

File: src/api/Jalasoft.TeamUp.Projects.API/local.settings.json (hunk 7 above)
Comment (pedro-lima-jalafund) [reply src/api/Jalasoft.TeamUp.Projects.API/local.settings.json:8]: This is for testing this api while resumes api is running on a different port on our machines.

**Review by cristian-chavez-jalafund — APPROVED** [review]

**Review by rodrigo-baldiviezo-jalafund — APPROVED** [review]

**Review by alejandro-castellon — APPROVED** [review]
Approved
